### PR TITLE
feat(gno): use `verbose` flag to print emitted events

### PR DESCRIPTION
**Pull request by r3v4s**
Closes half of #2007 

Add `print-events` flag to print emitted events in `gno -test`.

### Sample
```go
// event.gno
package event

import (
	"std"
)

func Hello() string {
	std.Emit("hello", "emit_key", "emit_value")

	return "hello_return"
}
```

```go
// event_test.gno
package event

import (
	"std"
	"testing"
)

func TestFunc(t *testing.T) {
	Hello()

	std.Emit("tEsTfUnC", "test key", "test value")
}

func TestOtherEvent(t *testing.T) {
	std.Emit("otherEvent", "test key", "test value")
}

```

### Ouput
image

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Review by moul — CHANGES_REQUESTED** [review]
Events printing should be as common as stdout. It should be the default on `gnokey maketx` and similar commands.

Regarding gnotest, I'm considering whether your approach is the best or if we should print nothing by default and only print everything (stdout and events) when using the `-v` option.

Blocking. We should wait for more feedback before making a decision.

**Comment by leohhhn** [comment]
Hey @moul, [here](https://github.com/gnolang/gno/issues/2007) is the issue with the discussion about this - TL;DR I am for adding `-print-events` apart from `-v`, since the events might clog up the test output if there is lots of them, making the output unreadable.

File: gnovm/cmd/gno/test.go [hunk 1]
```diff
@@ -500,10 +512,25 @@ func runTestFiles(
...
+		eval := m.Eval(res) // NOTE: verbose prints get here
+		if printEvents {
+			ctx := m.Context.(stdlibs.ExecContext)
+
+			events := ctx.EventLogger.Events()
+			for _, ev := range events[printedEvents:] {
+				printedEvents++
+				// XXX: print events with better formatting (e.g. JSON)
+				strEv := fmt.Sprint(ev)
+				io.ErrPrintfln("---       event: %s", colors.ColoredBytesOnlyAscii([]byte(strEv), colors.Magenta))
```
Comment (zivkovicmilos) [inline gnovm/cmd/gno/test.go:531]: Why is this an error print?

File: tm2/pkg/colors/colors.go [hunk 2]
```diff
@@ -107,3 +107,15 @@ func ColoredBytes(data []byte, textColor, bytesColor func(...interface{}) string
...
+
+func ColoredBytesOnlyAscii(data []byte, textColor func(...interface{}) string) string {
+	s := ""
+	for _, b := range data {
+		if 0x21 <= b && b < 0x7F {
+			s += textColor(string(b))
+		} else {
+			s += string(b)
+		}
+	}
```
Comment (zivkovicmilos) [inline tm2/pkg/colors/colors.go:119]:
```suggestion
	for _, b := range data {
		if 0x21 <= b && b < 0x7F {
			s += textColor(string(b))
			
			continue
		}
		
		s += string(b)
	}
```

File: tm2/pkg/colors/colors.go [hunk 3]
```diff
@@ -107,3 +107,15 @@ func ColoredBytes(data []byte, textColor, bytesColor func(...interface{}) string
 	}
 	return s
 }
+
+func ColoredBytesOnlyAscii(data []byte, textColor func(...interface{}) string) string {
+	s := ""
+	for _, b := range data {
+		if 0x21 <= b && b < 0x7F {
```
Comment (zivkovicmilos) [inline tm2/pkg/colors/colors.go:114]: What do you think about defining a method called `isASCII` that does this check on a single byte?

File: gnovm/cmd/gno/test.go [hunk 4]
```diff
@@ -500,10 +512,25 @@ func runTestFiles(
...
+	printedEvents := 0
+
 	for _, test := range testFuncs.Tests {
 		testFuncStr := fmt.Sprintf("%q", test.Name)
 
-		eval := m.Eval(gno.Call("runtest", testFuncStr))
+		res := gno.Call("runtest", testFuncStr)
+
+		eval := m.Eval(res) // NOTE: verbose prints get here
+		if printEvents {
```
Comment (zivkovicmilos) [inline gnovm/cmd/gno/test.go:523]: I agree with @moul, I think we should print these events if the `verbose` flag is enabled, and not hide this functionality behind an additional flag we need to now maintain

The reasoning is that events are now an integral part of the Gno development lifecycle, and not a "feature" we can disable

File: gnovm/cmd/gno/test.go [hunk 5]
```diff
@@ -500,10 +512,25 @@ func runTestFiles(
...
+			ctx := m.Context.(stdlibs.ExecContext)
+
+			events := ctx.EventLogger.Events()
+			for _, ev := range events[printedEvents:] {
+				printedEvents++
+				// XXX: print events with better formatting (e.g. JSON)
+				strEv := fmt.Sprint(ev)
```
Comment (zivkovicmilos) [inline gnovm/cmd/gno/test.go:530]: This can be inlined

File: gnovm/cmd/gno/test.go [hunk 6]
```diff
@@ -500,10 +512,25 @@ func runTestFiles(
...
 
-		eval := m.Eval(gno.Call("runtest", testFuncStr))
+		res := gno.Call("runtest", testFuncStr)
+
+		eval := m.Eval(res) // NOTE: verbose prints get here
+		if printEvents {
+			ctx := m.Context.(stdlibs.ExecContext)
+
+			events := ctx.EventLogger.Events()
+			for _, ev := range events[printedEvents:] {
```
Comment (zivkovicmilos) [inline gnovm/cmd/gno/test.go:527]: Are the events accumulated between different `gno.Call`s and `m.Eval`s, and that's why you're skipping previously printed events?

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
I've left a few comments that need addressing before we can move forward, otherwise it looks good 💯

File: gnovm/cmd/gno/test.go (hunk 1 above)
Comment (r3v4s) [reply gnovm/cmd/gno/test.go:531]: 99% of printing in that file uses error.

For example, metrics are printed using error
https://github.com/gnolang/gno/blob/9eeb61156d79208cd63c23bbe8c3a03806240f2b/gnovm/cmd/gno/test.go#L569

File: gnovm/cmd/gno/test.go (hunk 4 above)
Comment (r3v4s) [reply gnovm/cmd/gno/test.go:523]: > The reasoning is that events are now an integral part of the Gno development lifecycle

I totally understand, however for easier(or at least more readable) debugging flags needs to be separated.

File: gnovm/cmd/gno/test.go (hunk 6 above)
Comment (r3v4s) [reply gnovm/cmd/gno/test.go:527]: Yes it is.

File: gnovm/cmd/gno/test.go (hunk 5 above)
Comment (r3v4s) [reply gnovm/cmd/gno/test.go]: ed6f974918c907fb1d5b93aeaf12514383447aa0

File: tm2/pkg/colors/colors.go (hunk 3 above)
Comment (r3v4s) [reply tm2/pkg/colors/colors.go]: 875394f02faf82f6e2f4b1136d1be3ae3666f257

**Comment by moul** [comment]
If stdout is printed, events should also be printed if they exist. Currently, events should be considered as returned values of a contract, representing an outcome or consequence as the returned value (stdout). 

Let's wait for a clear example to determine if events are too spammy. Initially, we should consider reducing the number of events. If they overwhelm users, it makes the contract's usage more implicit, which we aim to avoid.

We should always print events when stdout is printed and maintain our usual rule: strive for straightforward, explicit, and simple communication without hiding things as soon as they become complex; let's just remove the complex.

**Review by zivkovicmilos — APPROVED** [review]

**Comment by zivkovicmilos** [comment]
@r3v4s 

Can you please check why the CI is failing and apply @moul's comment, so we can go ahead with merging this? 🙏

File: gnovm/cmd/gno/test.go (hunk 4 above)
Comment (r3v4s) [reply gnovm/cmd/gno/test.go]: 446eee7d57bd852f13faac818a7b56dd1436e1ba

**Comment by r3v4s** [comment]
@zivkovicmilos 
Removed new flag(`print-events) and applied existing flag `verbose` to print emitted events or not.

And CI fails on codecov against `ColoredBytesOnlyAscii` function, which I can't think of straightforward way to make it not to fail. Can you enlighten me?

**Comment by moul** [comment]
Closed in favor of #2975. But thank you @r3v4s!

**Pull request closed without merging**